native/packages/react-native-bpk-component-progress/src/BkpProgress.js

```diff
@@ -0,0 +1,129 @@
+/*
+ * Backpack - Skyscanner's Design System
+ *
+ * Copyright 2018 Skyscanner Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* @flow */
+
+import React from 'react';
+import PropTypes from 'prop-types';
+import { StyleSheet, View, Animated, Easing } from 'react-native';
+import {
+  animmationDurationBase,
+  spacingMd,
+} from 'bpk-tokens/tokens/base.react.native';
+
+import STYLES from './styles';
+
+const styles = StyleSheet.create({
+  container: {
+    height: spacingMd,
+    overflow: 'hidden',
+  },
+  fill: {
+    height: spacingMd,
+  },
+});
+
+type Props = {
+  max: number,
+  min: number,
+  value: number,
+  type: 'Default' | 'Bar',
+  trackStyle: StyleSheet,
+  fillStyle: StyleSheet,
+};
+
+type State = {
+  progress: Animated.Value,
+  width: number,
+};
+
+class ProgressBar extends React.Component<Props, State> {
+  static defaultProps: Object;
+
+  constructor(props: Props) {
+    super(props);
+    this.state = {
```

Comment:
I wouldn't expect this component to have any state - we should leave this to consumer land (i.e. like you've done in the `stories.js` container component).

Reply:
I can't see how I'd implement this without the state. From my understanding I need to save the animation in the state. Progress is the animation, the value prop is the actual progress, maybe I should change this names. 
Width is in the state because I need the computed width to do the animation, as I doesn't work with percentages, the "real" width still lives in the "user land" as they can provide it via the style prop.

Reply:
You cam avoid using React "state" by declaring it on the instance?

Example here: https://github.com/Skyscanner/backpack/blob/master/native/packages/react-native-bpk-component-animate-height/src/BpkAnimateHeight.js#L39

Technically this is still "state", but we want to avoid using reacts state api because it will involve unnecessary re-renders.

Reply:
I think it would work. I honestly don't think it's going to make any difference in this case, performance wise. This property is only set once in the constructor so no extra render involved. As for the width property the rerender is needed in my current implementation. I noticed in the example you've linked the `measure` api. I think I can use that for the width, not sure witch one is faster though.

Reply:
I tried with the measure api but I haven't found a way to do this without state because I need the width for the interpolate function. Even if I use the measure api I need to use state to trigger a rerender after getting the real width. The good things is, I tested and onLayout is only trigger after the first mount, or if the orientation/style changes. Other updates don't trigger it again.